File: student-web/src/api/apiClient.js
// ===== API client placeholder =====
// This module is the single integration boundary for the future FastAPI
// backend. Nothing in Stage 1 calls this for real; services currently talk
// to the mock/ layer instead. Once Stage 2 begins, services are switched to
// call this client, and no other file in the app needs to change.
//
//   Student UI -> Service -> [this file] -> FastAPI backend -> PostgreSQL

const BASE_URL = '/api'; // will become the deployed backend origin

let authToken = null;

export function setAuthToken(token) {
  authToken = token;
}

function buildHeaders(extra = {}) {
  const headers = { 'Content-Type': 'application/json', ...extra };
  if (authToken) headers.Authorization = `Bearer ${authToken}`;
  return headers;
}

/**
 * Thin wrapper around fetch(). Not used during Stage 1 — present so the
 * shape of the future integration is explicit and reviewable now.
 */
export async function apiRequest(path, { method = 'GET', body, headers } = {}) {
  const response = await fetch(`${BASE_URL}${path}`, {
    method,
    headers: buildHeaders(headers),
    body: body ? JSON.stringify(body) : undefined,
  });

  if (!response.ok) {
    const error = new Error(`API request failed: ${response.status}`);
    error.status = response.status;
    throw error;
  }

  if (response.status === 204) return null;
  return response.json();
}

export const apiClient = { get: (p) => apiRequest(p), post: (p, body) => apiRequest(p, { method: 'POST', body }), patch: (p, body) => apiRequest(p, { method: 'PATCH', body }) };

File: admin_desktop/api/api_client.py
"""
API CLIENT BOUNDARY (Stage 2 placeholder)
==========================================
This is the single centralized place that will eventually own all HTTP
communication with the FastAPI backend. No Tkinter UI code should ever
import `requests`/`httpx` directly, and no API URLs should be scattered
outside this file.

Stage 1: this class is NOT used by the running application. The service
layer (see services/) talks to mock/ modules instead. It exists now so the
integration boundary is explicit and so services can be swapped from
MockXService to ApiXService later without touching the UI.

Stage 2: services will construct requests through this client, which will
own:
    * base URL configuration
    * default headers / auth token attachment
    * timeout handling
    * centralized error handling & response parsing
    * retry policy (if any)
"""

from config.settings import API_BASE_URL, API_TIMEOUT_SECONDS


class ApiClientError(Exception):
    """Raised for any failed API call once real networking is implemented."""


class ApiClient:
    def __init__(self, base_url: str = API_BASE_URL, timeout: int = API_TIMEOUT_SECONDS):
        self.base_url = base_url
        self.timeout = timeout
        self._auth_token: str | None = None

    def set_auth_token(self, token: str) -> None:
        self._auth_token = token

    def clear_auth_token(self) -> None:
        self._auth_token = None

    # ------------------------------------------------------------------
    # The methods below are intentionally unimplemented in Stage 1.
    # Stage 2 will implement them using `requests`/`httpx`, attaching
    # `Authorization: Bearer <token>` headers from self._auth_token, and
    # raising ApiClientError on non-2xx responses.
    # ------------------------------------------------------------------

    def get(self, path: str, params: dict | None = None):
        raise NotImplementedError(
            "ApiClient.get() is a Stage 2 integration point. "
            f"Would call: GET {self.base_url}{path}"
        )

    def post(self, path: str, payload: dict | None = None):
        raise NotImplementedError(
            "ApiClient.post() is a Stage 2 integration point. "
            f"Would call: POST {self.base_url}{path}"
        )

    def put(self, path: str, payload: dict | None = None):
        raise NotImplementedError(
            "ApiClient.put() is a Stage 2 integration point. "
            f"Would call: PUT {self.base_url}{path}"
        )

    def delete(self, path: str):
        raise NotImplementedError(
            "ApiClient.delete() is a Stage 2 integration point. "
            f"Would call: DELETE {self.base_url}{path}"
        )


# Single shared instance the service layer will import in Stage 2, e.g.:
#   from api.api_client import api_client
api_client = ApiClient()
